File: 48_positional_encoding_timeseries/python/examples/trading_strategy.py
#!/usr/bin/env python3
"""
Trading Strategy with Positional Encoding Example

This example shows how to combine positional encodings with trading
strategies and backtesting for financial time series.
"""

import sys
import os
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import numpy as np
import torch
from datetime import datetime, timedelta

from positional_encoding import (
    SinusoidalPositionalEncoding,
    CalendarEncoding,
    MarketSessionEncoding,
)
from data import KlineData, prepare_features, create_sequences, train_test_split_time_series
from strategy import TradingStrategy, run_backtest, calculate_buy_and_hold, compare_strategies


def generate_synthetic_data(n_samples: int, seed: int = 42) -> tuple:
    """Generate synthetic price data for demonstration."""
    np.random.seed(seed)

    base_time = datetime(2024, 1, 1, 0, 0)
    timestamps = np.array([
        int((base_time + timedelta(hours=i)).timestamp())
        for i in range(n_samples)
    ])

    # Random walk with trend and volatility
    returns = np.random.normal(0.0001, 0.02, n_samples)
    prices = 45000 * np.cumprod(1 + returns)

    # Generate OHLCV
    opens = np.roll(prices, 1)
    opens[0] = 45000
    closes = prices
    highs = np.maximum(opens, closes) * (1 + np.abs(np.random.normal(0, 0.005, n_samples)))
    lows = np.minimum(opens, closes) * (1 - np.abs(np.random.normal(0, 0.005, n_samples)))
    volumes = np.random.uniform(500, 1500, n_samples)

    return timestamps, opens, highs, lows, closes, volumes, returns


def main():
    print("Trading Strategy Example")
    print("=" * 50)

    # 1. Generate or load data
    print("\n1. Data Preparation")
    print("-" * 40)

    n_samples = 2000
    timestamps, opens, highs, lows, closes, volumes, returns = generate_synthetic_data(n_samples)

    print(f"Generated {n_samples} hourly candles")
    print(f"Price range: ${closes.min():.2f} - ${closes.max():.2f}")
    print(f"Mean return: {returns.mean():.6f}")
    print(f"Std return: {returns.std():.6f}")

    # 2. Feature Engineering
    print("\n2. Feature Engineering")
    print("-" * 40)

    # Create KlineData object for prepare_features
    kline_data = KlineData(
        timestamp=timestamps,
        open=opens,
        high=highs,
        low=lows,
        close=closes,
        volume=volumes,
        symbol="SYNTH",
        interval="1h"
    )
    features_df = prepare_features(kline_data)
    print(f"Base features shape: {features_df.shape}")
    feature_cols = ['log_return', 'volatility', 'volume_change', 'momentum', 'rsi', 'range', 'vwap_deviation']
    features_numeric = features_df[feature_cols].values
    print(f"Features: {feature_cols}")

    # Add positional encoding
    d_model = 16
    n_features_len = len(features_df)
    pos_encoding = SinusoidalPositionalEncoding(d_model, n_features_len)
    x_pos = torch.zeros(1, n_features_len, d_model)
    pos_features = pos_encoding(x_pos).squeeze(0).detach().numpy()
    print(f"Position encoding shape: {pos_features.shape}")

    # Add calendar encoding - requires discrete features
    calendar = CalendarEncoding(d_model)
    dayofweek = torch.tensor(features_df['dayofweek'].values).unsqueeze(0)
    month = torch.tensor(features_df['month'].values - 1).unsqueeze(0)  # 0-indexed
    quarter = torch.tensor(features_df['quarter'].values).unsqueeze(0)
    hour = torch.tensor(features_df['hour'].values).unsqueeze(0)
    session_idx = torch.zeros_like(hour)  # placeholder
    cal_features = calendar(dayofweek, month, quarter, hour, session_idx).squeeze(0).detach().numpy()
    print(f"Calendar encoding shape: {cal_features.shape}")

    # Add market session encoding (crypto) - requires hour tensor
    session_enc = MarketSessionEncoding(d_model, market_type='crypto')
    session_features = session_enc(hour).squeeze(0).detach().numpy()
    print(f"Session encoding shape: {session_features.shape}")

    # Total features
    total_dim = features_numeric.shape[1] + pos_features.shape[1] + cal_features.shape[1] + session_features.shape[1]
    print(f"Total feature dimension: {total_dim}")

    # 3. Create Training Sequences
    print("\n3. Sequence Creation")
    print("-" * 40)

    sequence_length = 24  # 24-hour lookback
    target_length = 1     # 1-hour prediction

    X, y, seq_timestamps = create_sequences(
        features_df, feature_cols, target_col='log_return',
        seq_len=sequence_length, horizon=target_length
    )
    print(f"Created {len(X)} sequences")
    print(f"Sequence shape: {X.shape}")

    # Train/test split
    train_ratio = 0.8
    X_train, X_val, X_test, y_train, y_val, y_test = train_test_split_time_series(
        X, y, timestamps=None, train_ratio=train_ratio, val_ratio=0.1
    )
    print(f"Train: {len(X_train)} sequences, Val: {len(X_val)}, Test: {len(X_test)} sequences")

    # 4. Trading Strategy
    print("\n4. Trading Strategy")
    print("-" * 40)

    strategy = TradingStrategy(threshold=0.001, max_position=1.0, transaction_cost=0.001)

    print("Strategy configuration:")
    print(f"  Threshold: {strategy.threshold:.4f}")
    print(f"  Max Position: {strategy.max_position:.1f}")
    print(f"  Transaction Cost: {strategy.transaction_cost:.4f}")

    # Get returns from features_df for backtesting
    returns_for_backtest = features_df['log_return'].values

    # Generate signals from a simple moving average crossover (simulating model predictions)
    window = 24
    predictions = np.zeros(len(returns_for_backtest))
    for i in range(window, len(returns_for_backtest)):
        ma_short = returns_for_backtest[i-12:i].mean()
        ma_long = returns_for_backtest[i-24:i].mean()
        predictions[i] = ma_short - ma_long  # Momentum signal

    signals = strategy.generate_signals(predictions)
    positions = strategy.generate_position_sizes(signals)

    long_count = np.sum(signals > 0)
    short_count = np.sum(signals < 0)
    neutral_count = np.sum(signals == 0)

    print("\nSignal distribution:")
    print(f"  Long: {long_count} ({100.0 * long_count / len(signals):.1f}%)")
    print(f"  Short: {short_count} ({100.0 * short_count / len(signals):.1f}%)")
    print(f"  Neutral: {neutral_count} ({100.0 * neutral_count / len(signals):.1f}%)")

    # 5. Backtesting
    print("\n5. Backtesting")
    print("-" * 40)

    result = run_backtest(predictions, returns_for_backtest, strategy, periods_per_year=8760)

    print("\nStrategy Performance:")
    print(f"  Total Return: {result.total_return * 100:.2f}%")
    print(f"  Sharpe Ratio: {result.sharpe_ratio:.2f}")
    print(f"  Sortino Ratio: {result.sortino_ratio:.2f}")
    print(f"  Max Drawdown: {result.max_drawdown * 100:.2f}%")
    print(f"  Win Rate: {result.win_rate * 100:.2f}%")
    print(f"  Number of Trades: {result.n_trades}")

    # 6. Benchmark Comparison
    print("\n6. Benchmark Comparison")
    print("-" * 40)

    bh_result = calculate_buy_and_hold(returns_for_backtest, periods_per_year=8760)

    print("\nBuy & Hold Performance:")
    print(f"  Total Return: {bh_result.total_return * 100:.2f}%")
    print(f"  Sharpe Ratio: {bh_result.sharpe_ratio:.2f}")
    print(f"  Max Drawdown: {bh_result.max_drawdown * 100:.2f}%")

    print("\nStrategy vs Buy & Hold:")
    excess_return = result.total_return - bh_result.total_return
    sharpe_diff = result.sharpe_ratio - bh_result.sharpe_ratio
    print(f"  Excess Return: {excess_return * 100:.2f}%")
    print(f"  Sharpe Difference: {sharpe_diff:.2f}")

    # 7. Threshold Comparison
    print("\n7. Threshold Optimization")
    print("-" * 40)

    thresholds = [0.0005, 0.001, 0.002, 0.005]
    comparison = compare_strategies(predictions, returns_for_backtest, thresholds, periods_per_year=8760)

    print("\n| Threshold | Return   | Sharpe | MaxDD   | Trades |")
    print("|-----------|----------|--------|---------|--------|")
    for _, row in comparison.iterrows():
        print(f"| {row['threshold']:.4f}    | {row['total_return']*100:>7.2f}% | {row['sharpe_ratio']:>6.2f} | {row['max_drawdown']*100:>6.2f}% | {row['n_trades']:>6} |")

    # 8. Summary
    print("\n8. Integration with Positional Encoding")
    print("-" * 40)

    print("\nHow positional encoding helps trading strategies:")
    print()
    print("1. Sequence Position (Sinusoidal/RoPE):")
    print("   - Helps model understand temporal order of observations")
    print("   - Enables attention to focus on recent vs distant past")
    print()
    print("2. Calendar Features:")
    print("   - Captures day-of-week effects (Monday dip, Friday close)")
    print("   - Month-end rebalancing patterns")
    print("   - Seasonal/quarterly patterns")
    print()
    print("3. Market Session:")
    print("   - Regional activity patterns (Asia/Europe/Americas)")
    print("   - Session overlap volatility")
    print("   - Pre/post market behavior")
    print()
    print("Recommended Pipeline:")
    print("  1. Prepare base features (OHLCV + technical indicators)")
    print("  2. Add positional encoding (RoPE for transformers)")
    print("  3. Add calendar encoding (for time-aware predictions)")
    print("  4. Add market session encoding (for intraday)")
    print("  5. Train transformer model with combined features")
    print("  6. Generate predictions and signals")
    print("  7. Backtest with realistic transaction costs")
    print("  8. Optimize threshold and position sizing")


if __name__ == "__main__":
    main()

File: 48_positional_encoding_timeseries/rust/examples/trading_strategy.rs
//! Example: Trading Strategy with Positional Encoding
//!
//! This example shows how to combine positional encodings with trading
//! strategies and backtesting for financial time series.

use positional_encoding::{
    // Data
    generate_synthetic_data, FeaturePreparator, SequenceCreator, Candle,
    // Encodings
    CalendarEncoding, MarketSessionEncoding, MultiScaleTemporalEncoding, MarketType,
    SinusoidalEncoding, PositionalEncoding,
    // Strategy
    TradingStrategy, Signal, run_backtest, calculate_buy_and_hold, compare_strategies,
};

fn main() {
    println!("Trading Strategy Example");
    println!("========================\n");

    // 1. Generate or load data
    println!("1. Data Preparation");
    println!("-------------------");

    // Generate synthetic hourly data (1 year = 8760 hours)
    let candles = generate_synthetic_data(2000, 42);
    println!("Generated {} hourly candles", candles.len());

    // Extract price information
    let closes: Vec<f64> = candles.iter().map(|c| c.close).collect();
    let timestamps: Vec<i64> = candles.iter().map(|c| c.timestamp).collect();

    // Calculate returns
    let returns = FeaturePreparator::calculate_returns(&closes);
    println!("Calculated {} return observations", returns.len());

    // Show statistics
    let mean_return: f64 = returns.iter().sum::<f64>() / returns.len() as f64;
    let std_return: f64 = (returns.iter().map(|r| (r - mean_return).powi(2)).sum::<f64>()
        / returns.len() as f64).sqrt();
    println!("Mean return: {:.6}", mean_return);
    println!("Std return: {:.6}", std_return);

    // 2. Feature Engineering with Positional Encoding
    println!("\n2. Feature Engineering");
    println!("----------------------");

    // Prepare base features from candles
    let base_features = FeaturePreparator::prepare_features(&candles);
    println!("Base features shape: {:?}", base_features.shape());
    println!("Features: return, log_volume, range, body_ratio, direction, return_pct");

    // Add positional encoding
    let pos_encoding = SinusoidalEncoding::new(16, 2000);
    let positions: Vec<usize> = (0..candles.len()).collect();
    let pos_features = pos_encoding.encode(&positions);
    println!("Position encoding shape: {:?}", pos_features.shape());

    // Add calendar encoding
    let calendar = CalendarEncoding::new();
    let cal_features = calendar.encode_timestamps(&timestamps);
    println!("Calendar encoding shape: {:?}", cal_features.shape());

    // Add market session encoding
    let session = MarketSessionEncoding::new(MarketType::Crypto);
    let session_features = session.encode_timestamps(&timestamps);
    println!("Session encoding shape: {:?}", session_features.shape());

    // Total feature dimension
    let total_dim = base_features.ncols() + pos_features.ncols()
        + cal_features.ncols() + session_features.ncols();
    println!("Total feature dimension: {}", total_dim);

    // 3. Create Training Sequences
    println!("\n3. Sequence Creation");
    println!("--------------------");

    let targets: Vec<f64> = candles.iter().map(|c| c.return_pct()).collect();
    let creator = SequenceCreator::new(24, 1);  // 24-hour lookback, 1-hour prediction

    let (sequences, target_arrays, _) = creator.create_sequences(&base_features, &targets, &timestamps);
    println!("Created {} sequences", sequences.len());
    println!("Sequence shape: {:?}", sequences[0].shape());

    // Train/test split
    let (train_seq, test_seq) = SequenceCreator::train_test_split(&sequences, 0.8);
    let (train_targets, test_targets) = SequenceCreator::train_test_split(&target_arrays, 0.8);
    println!("Train: {} sequences, Test: {} sequences", train_seq.len(), test_seq.len());

    // 4. Trading Strategy
    println!("\n4. Trading Strategy");
    println!("-------------------");

    // Create a simple momentum strategy
    // (In practice, this would use model predictions)
    let strategy = TradingStrategy::new(0.001)
        .with_max_position(1.0)
        .with_transaction_cost(0.001);

    println!("Strategy configuration:");
    println!("  Threshold: {:.4}", strategy.threshold);
    println!("  Max Position: {:.1}", strategy.max_position);
    println!("  Transaction Cost: {:.4}", strategy.transaction_cost);

    // Generate signals from a simple moving average crossover
    // (simulating model predictions)
    let window = 24;
    let mut predictions = vec![0.0; returns.len()];
    for i in window..returns.len() {
        let ma_short: f64 = returns[i-12..i].iter().sum::<f64>() / 12.0;
        let ma_long: f64 = returns[i-24..i].iter().sum::<f64>() / 24.0;
        predictions[i] = ma_short - ma_long;  // Momentum signal
    }

    // Generate signals
    let signals = strategy.generate_signals(&predictions);
    let long_count = signals.iter().filter(|&&s| s == Signal::Long).count();
    let short_count = signals.iter().filter(|&&s| s == Signal::Short).count();
    let neutral_count = signals.iter().filter(|&&s| s == Signal::Neutral).count();

    println!("\nSignal distribution:");
    println!("  Long: {} ({:.1}%)", long_count, 100.0 * long_count as f64 / signals.len() as f64);
    println!("  Short: {} ({:.1}%)", short_count, 100.0 * short_count as f64 / signals.len() as f64);
    println!("  Neutral: {} ({:.1}%)", neutral_count, 100.0 * neutral_count as f64 / signals.len() as f64);

    // 5. Backtesting
    println!("\n5. Backtesting");
    println!("--------------");

    let backtest_result = run_backtest(&predictions, &returns, &strategy, 100000.0, 8760);

    println!("\nStrategy Performance:");
    println!("  Total Return: {:.2}%", backtest_result.total_return * 100.0);
    println!("  Annual Return: {:.2}%", backtest_result.metrics.annual_return * 100.0);
    println!("  Volatility: {:.2}%", backtest_result.metrics.volatility * 100.0);
    println!("  Sharpe Ratio: {:.2}", backtest_result.sharpe_ratio);
    println!("  Sortino Ratio: {:.2}", backtest_result.sortino_ratio);
    println!("  Max Drawdown: {:.2}%", backtest_result.max_drawdown * 100.0);
    println!("  Calmar Ratio: {:.2}", backtest_result.metrics.calmar_ratio);
    println!("  Win Rate: {:.2}%", backtest_result.win_rate * 100.0);
    println!("  Number of Trades: {}", backtest_result.n_trades);

    // 6. Benchmark Comparison
    println!("\n6. Benchmark Comparison");
    println!("-----------------------");

    let bh_result = calculate_buy_and_hold(&returns, 100000.0, 8760);

    println!("\nBuy & Hold Performance:");
    println!("  Total Return: {:.2}%", bh_result.total_return * 100.0);
    println!("  Sharpe Ratio: {:.2}", bh_result.sharpe_ratio);
    println!("  Max Drawdown: {:.2}%", bh_result.max_drawdown * 100.0);

    println!("\nStrategy vs Buy & Hold:");
    let excess_return = backtest_result.total_return - bh_result.total_return;
    let sharpe_diff = backtest_result.sharpe_ratio - bh_result.sharpe_ratio;
    println!("  Excess Return: {:.2}%", excess_return * 100.0);
    println!("  Sharpe Difference: {:.2}", sharpe_diff);

    // 7. Threshold Comparison
    println!("\n7. Threshold Optimization");
    println!("-------------------------");

    let thresholds = vec![0.0005, 0.001, 0.002, 0.005];
    let comparison = compare_strategies(&predictions, &returns, &thresholds, 8760);

    println!("\n| Threshold | Return   | Sharpe | MaxDD   | Trades |");
    println!("|-----------|----------|--------|---------|--------|");
    for (threshold, result) in comparison {
        println!("| {:.4}    | {:>7.2}% | {:>6.2} | {:>6.2}% | {:>6} |",
            threshold,
            result.total_return * 100.0,
            result.sharpe_ratio,
            result.max_drawdown * 100.0,
            result.n_trades
        );
    }

    // 8. Summary
    println!("\n8. Integration with Positional Encoding");
    println!("---------------------------------------");

    println!("\nHow positional encoding helps trading strategies:");
    println!();
    println!("1. Sequence Position (Sinusoidal/RoPE):");
    println!("   - Helps model understand temporal order of observations");
    println!("   - Enables attention to focus on recent vs distant past");
    println!();
    println!("2. Calendar Features:");
    println!("   - Captures day-of-week effects (Monday dip, Friday close)");
    println!("   - Month-end rebalancing patterns");
    println!("   - Seasonal/quarterly patterns");
    println!();
    println!("3. Market Session:");
    println!("   - Regional activity patterns (Asia/Europe/Americas)");
    println!("   - Session overlap volatility");
    println!("   - Pre/post market behavior");
    println!();
    println!("4. Multi-Scale Temporal:");
    println!("   - Combines all temporal features");
    println!("   - Captures patterns at hourly, daily, weekly scales");
    println!();

    println!("Recommended Pipeline:");
    println!("  1. Prepare base features (OHLCV + technical indicators)");
    println!("  2. Add positional encoding (RoPE for transformers)");
    println!("  3. Add calendar encoding (for time-aware predictions)");
    println!("  4. Add market session encoding (for intraday)");
    println!("  5. Train transformer model with combined features");
    println!("  6. Generate predictions and signals");
    println!("  7. Backtest with realistic transaction costs");
    println!("  8. Optimize threshold and position sizing");

    println!("\nExample completed!");
}
